### PR TITLE
Allow custom cache objects using callback modules

### PR DESCRIPTION
**Pull request by kocolosk**
This patch allows custom objects derived from design documents to be exposed in the `ddoc_cache`. A custom object is defined by a module exporting the `recover/1` function which accepts the name of a database and returns {ok, term()}. The term will be cached as if it were a design doc. For the sake of simplicity each custom cache object is evicted when any ddoc in the database changes.

Once a custom callback module has been defined the object associated with the module can be retrieved via ddoc_cache:open(DbName, Mod).

BugzID: 42707

### CODE REVIEW

File: src/ddoc_cache_opener.erl [hunk 1]
```diff
@@ -54,7 +54,7 @@
 }).
 
 -record(st, {
-    db_ddocs,
+    callback_modules = [],
```
Comment (kocolosk) [inline src/ddoc_cache_opener.erl]: As far as I can tell this field was vestigial, so I repurposed it.

File: src/ddoc_cache_opener.erl [hunk 2]
```diff
@@ -178,6 +178,10 @@ handle_cast({do_evict, DbName}, St) ->
 
 handle_cast({do_evict, DbName, DDocIds}, St) ->
     ets_lru:remove(?CACHE, {DbName, validation_funs}),
```
Comment (eiri) [inline src/ddoc_cache_opener.erl]: This one can be removed, it covered by the next step

**Comment by davisp** [comment]
+1 to squerge after pulling out the old validation_funs removal that @eiri pointed out.

**Comment by iilyak** [comment]
+1 after fixing @eiri's note

**Comment by kocolosk** [comment]
Squerging.

**Pull request merged**